### ISSUE TITLE
v0.3.13: doctor check passes while doctor health/read surfaces keep marking live WAL suspect

### ISSUE THREAD

**Issue by alexx-ftw** [Dicklesworthstone/mcp_agent_mail_rust#159]
## Summary

On `am 0.3.13`, the service can be back to operational with a passing full doctor check while the cheap health probe still reports the live mailbox as broken because the WAL sidecar is header-only/truncated. The running service then repeatedly marks live SQLite as suspect and reconstructs from the archive during normal reads.

This is either a false-positive/incomplete health probe after successful recovery, or `doctor check` is missing a live WAL-sidecar condition that `doctor health` considers fatal. In either case the two doctor commands disagree, and read surfaces continue to trigger noisy archive reconstruction even after the mailbox is usable.

## Environment

- Version: `am 0.3.13`
- Self-update: `Up to date (v0.3.13)`
- Binary: `$HOME/.local/bin/am`
- OS: CachyOS Linux, x86_64, kernel `7.0.12-1-cachyos`
- Service: user systemd service running `am serve-http --host 127.0.0.1 --port 8765 --no-tui --no-auth`
- Local recovery/workaround env:
  - `AM_ATC_ENABLED=false`
  - `AM_HEALTH_SWEEP_ENABLED=false`
  - `FSQLITE_CONCURRENT_MODE=false`

## Current observed state

The mailbox was reconstructed from the Git archive after earlier `database disk image is malformed` failures. After that recovery, the tool is usable: HTTP service is active, MCP send/read works, file reservation create/release works, and `am self-update --check` reports v0.3.13 as current.

`am doctor check` reports no failures and ends with `All checks passed`:

```text
Doctor check:
  Primary issue: [WARN] Skipped hot query-plan diagnostics: read-only doctor check avoids WAL-capable DB opens
  Secondary findings: 0 fail, 3 warn.
  Breakdown: 0 live incident(s), 0 archive hygiene, 4 environment
  --- Live Operational Checks ---
  [OK] db_file_sanity
  [OK] database
  [OK] pool_init
  [OK] storage_root_writable
  [OK] archive_db_parity
  [OK] foreign_key_integrity
  [OK] server_port
  [OK] server_http_health
  [OK] server_jsonrpc_health
  ...
  [WARN] wal_mode
  ...
All checks passed.
```

But `am doctor health` reports the live mailbox needs repair:

```text
fail: live mailbox needs repair: SQLite WAL sidecar is header-only/truncated at $HOME/.mcp_agent_mail_git_mailbox_repo/storage.sqlite3-wal (32 bytes); run `am doctor repair` to quarantine it before opening the database; next: am doctor repair --dry-run
```

The service journal also repeatedly reports the live DB as suspect and reconstructs from archive during normal reads:

```text
INFO live SQLite is suspect; read surfaces will prefer archive snapshots verdict_state=suspect durability_state=degraded_read_only
INFO database reconstruction from archive complete stats=reconstructed 15 projects, 244 agents, 712 messages (998 recipients), 0 parse errors
WARN execute_statement_dispatch: using in-memory fallback path while parity-cert mode is enabled backend_kind="mem" mode="parity_cert" strict_reject=false statement_kind="select" decision_reason="correlated_exists_fallback"
```

## Expected behavior

One of these should be true:

1. If a 32-byte/header-only WAL sidecar is an actual live integrity problem, `am doctor check` should report it consistently instead of ending with `All checks passed`.
2. If full doctor/server/archive parity prove the mailbox is healthy after reconstruction, `am doctor health` and read surfaces should not continue to mark live SQLite as suspect or trigger repeated archive reconstruction notices.

## Related issues

This looks related but not identical to the existing corruption/reconstruction issues:

- #156: v0.3.13 real on-disk corruption under swarm load / release request for Frankensqlite fixes
- #152: canonical-confirmed SQLite corruption plus non-convergent reconstruct loop under concurrent writers
- #151: closed residual spurious archive reconstruction after v0.3.13 integrity guard fix
- #127 and #119: older truncated-WAL startup/repair loops

I opened a new issue because this specific residual state is the mismatch between passing `doctor check`, failing `doctor health`, and repeated suspect/reconstruct logs after recovery and with the HTTP/MCP service usable.

## Notes

I have not attached the mailbox or full logs because they contain local multi-agent coordination data. The snippets above are redacted to replace the home path with `$HOME`; I can provide more redacted command output if useful.

**Comment by Dicklesworthstone** [comment]
Fixed on `main` by `55fe0e25` (2026-06-17) — *magic-aware WAL classification*. The classifier now distinguishes a valid 32-byte **frameless idle WAL** (correct header magic, zero committed frames → `IdleEmpty`, benign and openable as-is) from a genuine truncation/corruption artifact (a sub-32-byte partial header, or 32 bytes with an *invalid* magic). So `doctor health` no longer false-fails on the header-only WAL sidecar, and read surfaces stop triggering needless archive reconstruction after a clean recovery — resolving the `doctor check` vs `doctor health` disagreement you saw.

This landed after v0.3.13 and ships in the next release (0.3.14, currently pending the frankensqlite 0.1.12 publish — see #156). Leaving open until that release; reopen if it recurs afterward.

**Comment by bitjson** [comment]
Opus 4.8:

> **Comment on #159 — root cause of the read-surface `degraded_read_only` loop: `probe_sidecar_state` flags a benign "WAL-without-SHM" layout as `Suspect` (independent of WAL size), with the false-positive carve-outs disabled in `fast()` mode.**
> 
> Re: the note that this is fixed on `main` by `55fe0e25` (magic-aware WAL classification), shipping in 0.3.14 — that fix resolves the WAL-*classification* trigger (the `doctor health` "header-only/truncated WAL" message). But the **read-surface `degraded_read_only` loop has a second, distinct trigger that `55fe0e25` does not touch, and it still reproduces on current `main` @ `c8ee6968` (which already includes `55fe0e25`)** — so 0.3.14 won't fully resolve it. `55fe0e25` changed `wal_classify.rs` / `pool.rs` / `cli`, **not** `mailbox_verdict.rs`'s `probe_sidecar_state`, which is the actual read-surface trigger and keys on `-shm` **file existence**, not WAL classification:
> 
> ## Root cause
> 
> `probe_sidecar_state` (`crates/mcp-agent-mail-db/src/mailbox_verdict.rs:1110`) maps a "`-wal` present, `-shm` absent" sidecar layout to `MailboxState::Suspect`:
> 
> ```rust
> // mailbox_verdict.rs ~1130
> (false, true, false, _) => ProbeResult::warn_state(
>     "sidecar_state",
>     format!("WAL exists without SHM ({detail})"),
>     MailboxState::Suspect,
> ),
> ```
> 
> `Suspect` → `DurabilityState::DegradedReadOnly` (`mailbox_verdict.rs:311`), which forces read surfaces onto archive snapshots. The read/recovery consumers — `live_db_is_suspect` (`crates/mcp-agent-mail-tools/src/lib.rs:859`, emits the suspect log line) and `build_recovery_status` — use `VerdictOptions::fast()` (`mailbox_verdict.rs:782`: `skip_archive_count` + `skip_db_sanity_check`). In fast mode the two false-positive carve-outs cannot fire on a `sidecar_state` failure:
> - `verdict_prefers_archive_snapshot_reads_for_primary_read_surface` (`:700`) is gated on `only_db_sanity_failures` — but `db_sanity` is forced *passed* in fast mode and the failure is `sidecar_state`.
> - `recovery_verdict_is_archive_lag_only` requires `archive_drift == DbAhead` + all-failures==`archive_db_parity` — but `skip_archive_count` forces `archive_drift = Skipped` and the parity probe isn't run.
> 
> So a benign WAL-without-SHM yields perpetual `degraded_read_only` with no escape hatch.
> 
> ## Why this is a false positive (and why it's not WAL size)
> 
> `-shm` is a derived, rebuildable WAL index; SQLite/the engine recreate it on open and replay the WAL. WAL *trustworthiness* is decided from WAL **content** by `wal_classify` (`classify_wal_sidecar` / `wal_sidecar_is_truncation_artifact`, magic + checksum), never from `-shm`. The engine's own code agrees: `cleanup_empty_wal_sidecar` ("SQLite recreates the WAL on the next write"; it quarantines an orphan **SHM-without-WAL**, never WAL-without-SHM) and `doctor/fixers/wal_shm_sidecar_drift.rs` ("SQLite recreates both on open"; detect-only). Real WAL corruption is already caught by `wal_classify` + `db_sanity`/`integrity`, none of which need `-shm`.
> 
> The `sidecar_state` arm keys purely on **file existence**, so it fires regardless of WAL size or validity — at both a 32-byte and a 115 KB valid WAL. This is why #159's "header-only/truncated (32-byte) WAL" framing (which comes from a *different* path — the `doctor health` `wal_classify` truncation message) doesn't fully explain the read-path loop: the read-path verdict trips on `-shm` absence alone.
> 
> ## Self-perpetuating
> 
> Once degraded, reads are served from temp archive snapshots and the live DB is never opened in WAL mode → `-shm` is never recreated and the live `-wal` is never checkpointed by a normal connection → the next `fast()` verdict sees WAL-without-SHM again. On a busy mailbox the live DB is rarely held open continuously, so a `-wal` with no `-shm` is the common idle state and the false-positive fires constantly.
> 
> ## Deterministic reproduction
> 
> A fresh, healthy DB (no archive drift, `integrity_check` clean) goes `degraded_read_only` purely from a `-wal`-without-`-shm` layout:
> 
> ```bash
> ROOT=/tmp/am-sidecar-repro; PORT=8798
> rm -rf "$ROOT"; mkdir -p "$ROOT"
> export STORAGE_ROOT="$ROOT" DATABASE_URL="sqlite:///$ROOT/storage.sqlite3"
> mcp(){ curl -s -m8 "http://127.0.0.1:$PORT/mcp/" -H 'Content-Type: application/json' \
>   -H 'Accept: application/json, text/event-stream' -d "$1"; }
> 
> am migrate
> # 1. start + one write to create a real WAL, then SIGKILL (leaves -wal, drops -shm)
> am serve-http --no-tui --no-auth --host 127.0.0.1 --port "$PORT" --path /mcp/ & sleep 6
> mcp '{"jsonrpc":"2.0","id":1,"method":"tools/call","params":{"name":"ensure_project","arguments":{"project_key":"/tmp/p","human_key":"p"}}}'
> mcp '{"jsonrpc":"2.0","id":2,"method":"tools/call","params":{"name":"register_agent","arguments":{"project_key":"/tmp/p","agent_name":"A","model":"m","program":"x","task_description":"t"}}}'
> kill -9 "$(ss -ltnp "sport = :$PORT" | grep -oE 'pid=[0-9]+' | head -1 | cut -d= -f2)"; sleep 2
> rm -f "$ROOT/storage.sqlite3-shm"          # ensure SHM absent
> ls "$ROOT"/storage.sqlite3-*               # -> -wal present, no -shm
> 
> # 2. restart on the WAL-without-SHM layout and check the verdict
> am serve-http --no-tui --no-auth --host 127.0.0.1 --port "$PORT" --path /mcp/ & sleep 8
> mcp '{"jsonrpc":"2.0","id":3,"method":"tools/call","params":{"name":"health_check","arguments":{}}}'
> # -> recovery.mode == "degraded_read_only"   (health_level green)
> am doctor check                            # -> "All checks passed" / integrity clean
> ```
> 
> Expected: `recovery.mode: degraded_read_only` **while** `am doctor check`/`integrity_check` is clean — a false positive driven solely by `-shm` absence.
> 
> ## Suggested fix
> 
> Gate the `(false, true, false, _)` arm (`mailbox_verdict.rs:1130-1134`) on actual WAL damage rather than mere `-shm` absence: treat a benign/idle/valid-frame WAL with a missing `-shm` as `ok` (or a non-impacting `info`), and only return `Suspect` when `wal_classify` reports a truncated/malformed WAL (`wal_sidecar_is_truncation_artifact` / `classify_wal_sidecar(...).state` damaged). `probe_sidecar_state` is currently path-free (`:1110`), so either thread the db path in or extend `MailboxSidecarState`/`inspect_mailbox_sidecar_state` (`pool.rs:~5919`) to carry the `wal_classify` verdict. Keep the existing SHM-without-WAL and non-file/`live_sidecars` arms (those are real signals). Do **not** gate on `live_sidecars` alone — a non-empty WAL sets `live_sidecars=true` yet still trips the existence arm. No existing test pins the buggy arm (the `probe_sidecar_state` tests cover the rollback-journal and non-file arms only), so this is additive; add a positive test that a benign WAL-without-SHM → non-`Suspect`.
> 
> Optionally also harden the read-path: have a fast-mode `sidecar_state`-only Suspect be eligible for the same archive-snapshot-suppression carve-out that `db_sanity`-only failures get, when the primary read path is healthy.
> 
> ## Related (filed separately)
> 
> There is a second, distinct false-positive in this verdict subsystem (filed as #163): `probe_archive_drift` maps **any** `DbAhead` → `Suspect` with no threshold (`mailbox_verdict.rs:1348`) on the non-fast verdict (`am robot health`/doctor). It interacts with this bug: a co-occurring `sidecar_state` failure (above) also **defeats** that bug's DB-ahead suppression carve-out (`mailbox_verdict.rs:676`, which requires `archive_db_parity` to be the sole failure). Fixing the `sidecar_state` false-positive here therefore also stops it from masking that carve-out.
> 
> ## Severity / scope
> 
> Non-destructive (integrity stays clean; data intact; writes work) but it pins long-running `serve-http` daemons in `degraded_read_only`, forcing an archive-snapshot rebuild on every read (latency + reconstruct churn) and emptying lexical `search_messages` (the snapshot's Search V3 index isn't backfilled — filed as #162). On a busy mailbox the trigger condition (idle `-wal` with no `-shm`) is the norm, so the false-positive is near-constant.

**Comment by Dicklesworthstone** [comment]
Fixed on `main` by two complementary commits:

- `55fe0e25` (magic-aware WAL classification) fixes the `doctor health` / `doctor repair` side: the "needs repair: SQLite WAL sidecar is header-only/truncated (32 bytes)" gate now routes through `wal_classify::wal_sidecar_is_truncation_artifact`, which treats a *valid* 32-byte frameless idle WAL (correct header magic, zero committed frames) as benign `IdleEmpty` and only flags a genuine truncation/garbage-magic artifact. So `doctor health` no longer false-fails on the header-only WAL that `doctor check` (and the live server) consider healthy — resolving the `doctor check` vs `doctor health` disagreement.
- `b7531127` fixes the runtime read-surface loop @bitjson identified: the durability `probe_sidecar_state` is now magic-aware for the WAL-without-SHM layout — a benign valid header-only WAL no longer escalates to `Suspect`/`degraded_read_only`, so read surfaces stop triggering needless archive reconstruction after a clean recovery. Tests: `probe_sidecar_state_benign_header_only_wal_without_shm_is_not_suspect` / `..._garbage_..._is_suspect`.

Reopen if the disagreement recurs after the next release.
